### ISSUE TITLE
keras should make sure outputs for cost function are sane

### ISSUE THREAD

**Issue by isaacgerg** [fchollet/keras#8460]
keras should check to ensure the output vector size is appropriate for the cost function.  For example, you can pass binary_crossentropy as a cost function with outputs that are of shape [batchsize, 28,28,10].  Keras is happy to compute with this setup though I'm not sure what exactly is going on.  Probably keras should tell me to stop being dumb and move to categorical_crossentorpy.

**Comment by fchollet** [comment]
> you can pass binary_crossentropy as a cost function with outputs that are of shape [batchsize, 28,28,10]

This is a valid use case (you'd use it in image segmentation with overlapping classes, for instance), so it should not be blocked.

We do minimal output/loss compatibility sanity checking (see `training.py`). If you have more ideas for new checks, please open a PR.

**Comment by isaacgerg** [comment]
@fchollet Thanks.  I didn't know feeding 10 dimensional vectors to binary_crossentropy would ever make sense.
